cylc/flow/dbstatecheck.py

```diff
@@ -151,36 +203,69 @@ def workflow_state_query(
                 stmt_args.append(state)
                 stmt_frags.append("status==?")
             stmt_wheres.append("(" + (" OR ").join(stmt_frags) + ")")
+
         if stmt_wheres:
-            stmt += " where " + (" AND ").join(stmt_wheres)
+            stmt += "WHERE\n    " + (" AND ").join(stmt_wheres)
+
+        if status:
+            stmt += dedent("""
+                ORDER BY
+                    submit_num
+            """)
+
+        # idx_status = 2
+        # idx_flow = 3
+
+        # TODO option to list outputs instead of status
+        # TODO if flows not specified, use the most recent for the task
 
         res = []
         for row in self.conn.execute(stmt, stmt_args):
-            if not all(v is None for v in row):
-                res.append(list(row))
+            if row[2] is None:
+                # status can be None in Cylc 7 DBs
+                continue
+            if not self.back_compat_mode:
+                flow_nums = deserialise(row[3])
+                if flow_num is not None and flow_num not in flow_nums:
+                    continue
+            res.append(list(row))
 
-        return res
+        if output:
+            # Replace res with a task-states like result,
+            # [[foo, 2032, output], [foo, 2033, output]]
+            if self.back_compat_mode:
+                # Cylc 7 DB: list of {output: message}
+                results = [
+                    [item[0], item[1], output]
+                    for item in res
+                    if output in deserialise(item[2]).values()
```

Comment:
For better readability 
```suggestion
                    [name, cycle, output]
                    for name, cycle, outputs_str in res
                    if output in deserialise(outputs_str).values()
```